kernel/mmu.c

```diff
@@ -1374,19 +1391,23 @@ int k_mem_page_frame_evict(uintptr_t phys)
 	}
 
 #ifdef CONFIG_DEMAND_PAGING_ALLOW_IRQ
-	irq_unlock(key);
+	k_spin_unlock(&z_mm_lock, key);
```

Comment:
Am I correct in thinking that  CONFIG_DEMAND_PAGING_ALLOW_IRQ can not be used in the SMP case? In the UP case a reschedule is prevented from occurring during the call to do_backing_store_page_out() by either the irq/spin lock or k_sched_lock() depending upon whether CONFIG_DEMAND_PAGING_ALLOW_IRQ is enabled. 

I may be wrong, but I am presently assuming that the reschedule prevention properties are why k_sched_lock() was chosen for UP. However, in the SMP case, the k_mutex_lock() does not prevent any rescheduling on that cpu if CONFIG_DEMAND_PAGING_ALLOW_IRQ is enabled. If SMP and allow IRQ do not mix, then the Kconfig for CONFIG_DEMAND_PAGING_ALLOW_IRQ should be updated.

I have some knowledge gaps here, so I can not say for certain what the desired behavior should be so I am throwing it out there.